### ISSUE TITLE
addmv bfloat16 accuracy issues on cpu

### ISSUE THREAD

**Issue by AnthonyBarbier** [pytorch/pytorch#147860]
### 🐛 Describe the bug

Significant inaccuracy when using addmv composite opinstead of individual ops with bf16 inputs :

```python
import torch

b = torch.tensor([5.8438, -6.3125], dtype=torch.bfloat16)
A = torch.tensor(
    [
        [4.3125, -6.4375, -3.8125, 4.3125, 8.2500, -5.4062, -3.2656, -5.4688, 6.1562, -2.9062],
        [7.0000, 0.7617, 5.1875, -4.9375, 4.5625, -6.5938, -0.9023, 6.4375, -2.4219, -3.3906],
    ],
    dtype=torch.bfloat16,
)

x = torch.tensor(
    [5.6875, -8.0000, 8.0625, 5.0000, -4.7812, 7.4375, -1.4766, 3.2344, 1.9688, -3.3125],
    dtype=torch.bfloat16,
)

c = b + (A @ x)
print(f"b + (A @ x) = {c}")
out = torch.addmv(b, A, x, alpha=1, beta=0)
print(f"addmv(beta=0) + b bfloat {out+b}")
print(f"addmv(beta=0) + b fp32 {(out.float()+b.float()).bfloat16()}")
out = torch.addmv(b, A, x, alpha=1, beta=1)
print(f"addmv bfloat {out}")
```

Output on x86 with https://download.pytorch.org/whl/nightly/cpu/torch-2.7.0.dev20250224%2Bcpu-cp313-cp313-manylinux_2_28_x86_64.whl

```
b + (A @ x) = tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv(beta=0) + b bfloat tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv(beta=0) + b fp32 tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv bfloat tensor([1.6562, 2.1562], dtype=torch.bfloat16)
```

Output on Graviton 3 with https://download.pytorch.org/whl/nightly/cpu/torch-2.7.0.dev20250224%2Bcpu-cp310-cp310-manylinux_2_28_aarch64.whl
```
b + (A @ x) = tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv(beta=0) + b bfloat tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv(beta=0) + b fp32 tensor([1.9219, 2.3125], dtype=torch.bfloat16)
addmv bfloat tensor([1.6562, 2.1562], dtype=torch.bfloat16)
```

It doesn't seem to be an accumulation issue as it only affects the "add" part.

I thought maybe it was an issue to do with the accumulator being fp32 which is why I tried to manually upcast the operands for the add but that didn't change anything.


### Versions

x86 Environment:
```
Collecting environment information...
PyTorch version: 2.7.0.dev20250224+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: EndeavourOS Linux (x86_64)
GCC version: (GCC) 14.2.1 20250207
Clang version: Could not collect
CMake version: version 3.31.5
Libc version: glibc-2.41

Python version: 3.13.2 (main, Feb  5 2025, 08:05:21) [GCC 14.2.1 20250128] (64-bit runtime)
Python platform: Linux-6.13.4-arch1-1-x86_64-with-glibc2.41
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Ti
Nvidia driver version: 570.86.16
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12700F
CPU family:                           6
Model:                                151
Thread(s) per core:                   2
Core(s) per socket:                   12
Socket(s):                            1
Stepping:                             2
CPU(s) scaling MHz:                   24%
CPU max MHz:                          4900.0000
CPU min MHz:                          800.0000
BogoMIPS:                             4224.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            512 KiB (12 instances)
L1i cache:                            512 KiB (12 instances)
L2 cache:                             12 MiB (9 instances)
L3 cache:                             25 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.7.0.dev20250224+cpu
[pip3] torchaudio==2.6.0.dev20250224+cpu
[pip3] torchvision==0.22.0.dev20250224+cpu
[conda] Could not collect
```

Aarch64 environment:

```
Collecting environment information...                                                                                                                                                                                                                                         
PyTorch version: 2.7.0.dev20250224+cpu                                                                                                                                                                                                                                        
Is debug build: False                                                                                                                                                                                                                                                         
CUDA used to build PyTorch: None   
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-1021-aws-aarch64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         aarch64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            ARM
Model:                                1
Thread(s) per core:                   1
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             r1p1
BogoMIPS:                             2100.00
Flags:                                fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm ssbs paca pacg dcpodp svei8mm svebf16 i8mm bf16 dgh rng
L1d cache:                            1 MiB (16 instances)
L1i cache:                            1 MiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             32 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; CSV2, BHB
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] torch==2.7.0.dev20250224+cpu
[pip3] torchaudio==2.6.0.dev20250224
[pip3] torchvision==0.22.0.dev20250224
[conda] Could not collect
```

**Comment by malfet** [comment]
Grabbing for myself to investigate(as I've touched addmv in the past) May be it has something to do with performing computations in bfloat16 instead of upcasting it to float32?

**Comment by f2013519** [comment]
I encountered a similar issue with dot as well.

The issue is that there are different code paths taken for beta=0 (optimized) and beta=1. beta=0 works as computations are done in float32 to preserve accuracy. When beta = 1, the current code performs does the computation in bfloat16 instead of float32. 

I have fixed both in #152676 , PTAL @malfet
